### PR TITLE
time sync decoding for asr

### PR DESCRIPTION
**Pull request by brianyan918**
This PR supports time synchronous decoding for ASR.

Librispeech_100 results:
### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|beam20_ctc0.3/dev_clean|2703|54402|94.5|5.1|0.4|0.8|6.3|56.3|
|beam20_ctc0.3/dev_other|2864|50948|84.6|13.9|1.5|2.1|17.4|79.9|
|beam20_ctc0.3/test_clean|2620|52576|94.3|5.3|0.4|0.8|6.5|57.0|
|beam20_ctc0.3/test_other|2939|52343|84.7|13.7|1.6|2.0|17.3|81.6|
|timesync_beam20_ctc0.3/dev_clean|2703|54402|94.4|5.1|0.5|0.7|6.3|56.6|
|timesync_beam20_ctc0.3/dev_other|2864|50948|83.9|13.4|2.7|1.8|17.8|80.3|
|timesync_beam20_ctc0.3/test_clean|2620|52576|94.3|5.2|0.5|0.7|6.5|57.3|
|timesync_beam20_ctc0.3/test_other|2939|52343|84.1|13.4|2.4|1.8|17.7|82.2|

### RTF 
image

Note on RTF:
* I have not yet implemented batching for the attentional decoder's fwd pass. This means that compared to the batchified label synchronous algorithm, this time synchronous algorithm is currently slower.
* This gap is widened on GPU inference, as shown above.

### CODE REVIEW

File: espnet2/bin/asr_inference.py [hunk 1]
```diff
@@ -229,39 +231,54 @@ def __init__(
...
-                            "BatchBeamSearchOnlineSim implementation is selected."
-                        )
+            if time_sync:
+                beam_search = TimeSyncBeamSearch(
+                    beam_size=beam_size,
+                    ctc=asr_model.ctc,
+                    sos=asr_model.sos,
+                    decoder=asr_model.decoder,
+                    lm=scorers["lm"] if "lm" in scorers else None,
+                    ctc_weight=ctc_weight,
```
Comment (sw005320) [inline espnet2/bin/asr_inference.py:242]: Can you make the weight and scores to be the same interface as the normal BeamSearch?

File: espnet2/bin/asr_inference.py [hunk 2]
```diff
@@ -229,39 +231,54 @@ def __init__(
...
-                    if streaming:
-                        beam_search.__class__ = BatchBeamSearchOnlineSim
-                        beam_search.set_streaming_config(asr_train_config)
-                        logging.info(
-                            "BatchBeamSearchOnlineSim implementation is selected."
-                        )
+            if time_sync:
+                beam_search = TimeSyncBeamSearch(
+                    beam_size=beam_size,
+                    ctc=asr_model.ctc,
```
Comment (sw005320) [inline espnet2/bin/asr_inference.py:238]: Can you add some error handling when there is no ctc?

File: espnet/nets/time_sync_beam_search.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        pre_beam_threshold = np.sort(p_ctc)[-self.pre_beam_size]
+        cands = set(np.where(p_ctc >= pre_beam_threshold)[0])
+        if len(cands) == 0:
+            cands = {np.argmax(p_ctc)}
+        new_hyps = set()
+        ctc_score_dp_next = defaultdict(
+            lambda: (float("-inf"), float("-inf"))
+        )  # (p_nb, p_b)
+        tmp = []
+        for l in hyps:
```
Comment (sw005320) [inline espnet/nets/time_sync_beam_search.py:160]: I recommend you add some `logging.debug` functions of how the hypothesis grows in each time step.

File: espnet/nets/time_sync_beam_search.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        penalty: float = 1.0,
+        blank: int = 0,
+        force_lid: bool = False,
+        temp: float = 1.0,
+    ):
+        """
+        beam_size: num hyps
+        sos: sos index
+        ctc: CTC module
+        pre_beam_ratio: pre_beam_ratio * beam_size = pre_beam
```
Comment (sw005320) [inline espnet/nets/time_sync_beam_search.py:60]: It is better to have more explanations of pre_beam

**Review by sw005320 — COMMENTED** [review]
- can you add the result in `egs2/librispeech_100/asr1/README.md`?
- does this support the GPU inference? (if so, please add some comments)
- add some test functions. Please check some examples in the normal beam search

**Comment by sw005320** [comment]
@YosukeHiguchi, could you also review this PR?

File: espnet2/bin/asr_inference.py (hunk 2 above)
Comment (brianyan918) [reply espnet2/bin/asr_inference.py]: done

File: espnet/nets/time_sync_beam_search.py (hunk 4 above)
Comment (brianyan918) [reply espnet/nets/time_sync_beam_search.py:57]: done

**Comment by sw005320** [comment]
> @sw005320 I think this passed CI except for an unrelated test

I'm working on it https://github.com/espnet/espnet/pull/4826

**Comment by sw005320** [comment]
Can you add a unit test?

**Comment by brianyan918** [comment]
> Can you add a unit test?

Sorry for the delay. I added a unit test; it is very basic and just checks that the beam search produces some result. I do not compare to any hand computed values.

**Review by sw005320 — COMMENTED** [review]
I have two comments

- naming: by following `beam_search.py` and `beam_search_transducer.py,` it would be better to name it as `beam_search_timesync.py.` Please also change the other names (e.g., the class name `TimeSyncBeamSearch` --> `BeamSearchTimeSync` accordingly.
- I'm thinking about where is the best place to put this function. We generally want to add a new function under the `espnet2` directory (e.g., `espnet2/asr`). But at the same time, putting it in its current place also makes sense in terms of putting all beam search-related functions in the same directory. I think we can leave it based on your idea, but I just added this remark for the future beam search design.

**Comment by brianyan918** [comment]
> 

Thanks @sw005320. This is fixed now.

**Comment by sw005320** [comment]
Thanks, @brianyan918 !

**Pull request merged**